content/post/codesent.md

```diff
@@ -0,0 +1,381 @@
+quieries on it.
+
+I need to say a few words about Tensorflow before explaining the comments classification model.
+
+{{% caption src="/post/codesent/tf_black.svg" %}}
+Tensorflow graph scheme.
+{{% /caption %}}
+
+The core of Tensorflow is the computational graph. It contains two types of nodes: tensors and operations.
+"Tensor" sounds a bit scientific; it is actually just a array with numbers. So buffers are specified
+as the input, they can represent the state of the graph which can mutate and there are output ones
+which we read in the end. Operations transform tensors according to the control flow rules.
+
+{{% caption src="https://www.tensorflow.org/images/tensors_flowing.gif" %}}
+Tensor flow. Taken from the official web site.
+{{% /caption %}}
+
+The rest is the layers of abstraction on top of the graph to simplify the typical ML solutions:
+neural networks, graph routines, visualization, etc. The key point is: Tensorflow abstracts the
+graph execution from the actual hardware where it executes. Tensors may be big and operations on them
+benefit from various accelerators such as GPUs, and Tensorflow allows to pin each operation to the
+supported execution device.
+
+Each ML model goes through the training phase, either supervised or not. The graph state tensors change
+while the model is trained. Here is the intended way to train and use a typical neural network:
+
+1. Debug the model locally: make sure it converges and there are no implementation faults. We debug on [e-GPUs](https://egpu.io/) at source{d}.
+2. Train the model on as much data as you have on a GPU cluster
+3. Optionally apply a metaparameter optimization algorithm to reach the best metrics in (2)
+4. Export the trained graph in "GraphDef" format (Protocol Buffers)
+5. Distribute it in an embedded or a pluggable fashion - that is, inside the application binary or separately
+6. Apply it ("infer") with any programming language
+
+In our case, I used Python API to Tensorflow on a dedicated machine with 4 GPUs running two experiments
+in parallel for training and Go API and plain CPU for inference. It is impossible to train Tensorflow
+models using the current Go bindings, but even if it was possible, I would still not do that. From my
+experience, scripting languages are superior for doing research where you rewrite code every now and then
+and need the full expressiveness and flexibility to achieve the result as soon as possible.
+At the same time Go fits the inference scenario well: we put everything into the single executable
+without dependencies and it just works. Everybody who deployed models in Python should understand
+the pain.
+
+Have you heard about [neugram](https://neugram.io/), a scripting language on top of Go runtime?
+It was created by David Crawshaw, the author of `pkg/plugin`. Neugram is still in it's early ages
+and has childhood problems, but I really like it. Who knows, maybe one day we will train Tensorflow
```

Comment:
s/childhood problems/youth related issues/ ?

Reply:
Need to remember this phrase instead of copying shit from Russian, thanks!